### PR TITLE
Charts

### PR DESCRIPTION
**Pull request by Oleksandr-Adamchuk**
**CHECK LIST**
- [x] specs added
- [x] specs passed
- [ ] code coverage >90%
- [x] DB indexes added if needed
- [x] seed.rb file is updated if needed
- [ ] PR is reviewed manually again (to make sure you have 100% ready code)
- [x] PR doesn't have conflicts with main branch
- [ ] I've checked new feature as logged in and logged out user if needed
- [ ] "let!" is used in specs
- [x] screenshot of rails_best_practices and other tools added
- [x] screenshot of feature is added if task has UI

1. Added Analytics Controller for Dashboard analytics section.
2. Added Chartkick gem, made charts for Admin with different data representation.
3. Create custom controller to track users logins.
4. Made charts to load asynchronously.
5. Integrated Code Climate.

dashboard

analytics top

analytics bottom

reek

### CODE REVIEW

File: app/controllers/account/analytics_controller.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+class Account::AnalyticsController < ApplicationController
+  before_action :authenticate_user!
+  layout 'account'
+
+  def index
+    redirect_to root_path unless current_user.admin?
```
Comment (igorkasyanchuk) [inline app/controllers/account/analytics_controller.rb]: use before_action ....
we have already something for admins

File: app/controllers/charts_controller.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
+class ChartsController < ApplicationController
+  include ApplicationHelper
```
Comment (igorkasyanchuk) [inline app/controllers/charts_controller.rb]: where is before_action? controller is 100% accessible

File: app/helpers/application_helper.rb [hunk 3]
```diff
@@ -10,4 +10,8 @@ def resource
   def devise_mapping
     @devise_mapping ||= Devise.mappings[:user]
   end
+
+  def three_month(user)
```
Comment (igorkasyanchuk) [inline app/helpers/application_helper.rb]: rework it. pass number of months to method. rename method.

and if you passing "user" => Company - it's a sample of bad naming for variable. Must be more abstract

File: config/routes.rb [hunk 4]
```diff
@@ -17,9 +17,17 @@
     end
     resources :tenants, only: %i[show index]
     resource :my_tenant, only: %i[show edit update]
+    resources :analytics, only: %i[index]
     resources :users do
       post :impersonate, on: :member
       post :stop_impersonating, on: :collection
     end
   end
+  namespace :charts do
```
Comment (igorkasyanchuk) [inline config/routes.rb]: charts must be in namespace "account"

File: app/controllers/account/analytics_controller.rb (hunk 1 above)
Comment (Oleksandr-Adamchuk) [reply app/controllers/account/analytics_controller.rb]: Done.

File: app/controllers/charts_controller.rb (hunk 2 above)
Comment (Oleksandr-Adamchuk) [reply app/controllers/charts_controller.rb]: Done.

File: app/helpers/application_helper.rb (hunk 3 above)
Comment (Oleksandr-Adamchuk) [reply app/helpers/application_helper.rb]: Done.

File: config/routes.rb (hunk 4 above)
Comment (Oleksandr-Adamchuk) [reply config/routes.rb]: Done.

File: app/views/account/analytics/index.html.slim [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+      download: true, xtitle: "Time", ytitle: "Companies count"
+  .row
+    .col-md-6
+      = column_chart account_charts_new_users_path, legend: true, label: "Users registraion per months",
+      download: true, xtitle: "Months", ytitle: "User count"
+    .col-md-6
+      = area_chart account_charts_new_companies_path, legend: true, label: "Company analytics", colors: ["#b00"],
+      download: true, xtitle: "Companies count group by day of a week", ytitle: "Company count"
+  .row
+    .col-md-6
```
Comment (romaslmd) [inline app/views/account/analytics/index.html.slim]: not necessary

File: config/routes.rb [hunk 6]
```diff
@@ -17,9 +17,17 @@
...
     resources :users do
       post :impersonate, on: :member
       post :stop_impersonating, on: :collection
     end
+    namespace :charts do
+        get 'new-users-by-week'
+        get 'new-companies-by-week'
+        get 'new-users'
+        get 'new_companies'
+        get 'users-login'
```
Comment (romaslmd) [inline config/routes.rb]: these doesn't look good.. at all

File: app/controllers/account/charts_controller.rb [hunk 7]
```diff
@@ -0,0 +1,25 @@
+class Account::ChartsController < ApplicationController
+  before_action :authorize_admin!
+  def users_by_week
+    chart_data = current_tenant.users
+    render json: chart_data.group_by_week(:created_at, range: time_range).count
+  end
+
+  def companies_by_week
```
Comment (romaslmd) [inline app/controllers/account/charts_controller.rb]: such suffixes are not relly good, as you may change chart to show monthly or yearly

File: app/controllers/account/charts_controller.rb [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    chart_data = current_tenant.companies
+    render json: chart_data.group_by_week('companies.created_at', range: time_range).count
+  end
+
+  def users_by_month
+    render json: current_tenant.users.group_by_month_of_year(:created_at).count
+  end
+
+  def users_logins
+    users_list = Visit.where(tenant_id: current_tenant.id)
```
Comment (igorkasyanchuk) [inline app/controllers/account/charts_controller.rb]: use relations from current_tenant

File: spec/controllers/charts_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,22 @@
+require 'rails_helper'
+
+RSpec.describe Account::ChartsController, type: :controller do
+  describe "GET requests to actions" do
+    it "returns a 200 status code" do
+      get :users_by_week
+      expect(response).to have_http_status(302)
```
Comment (igorkasyanchuk) [inline spec/controllers/charts_controller_spec.rb]: test for 302? this is a joke? what are you testing here?

File: db/seeds.rb [hunk 10]
```diff
@@ -19,3 +19,25 @@
...
+             domain: Faker::Internet.domain_name,
+             created_at: rand(3.month.ago..Time.now),
+             user_id: Faker::Number.between(1, 200)
+            })
+            User.create({
+             first_name: Faker::Name.first_name,
+             last_name: Faker::Name.last_name,
+             email: Faker::Internet.unique.email,
+             password: User::DEFAULT_PASSWORD,
+             tenant_id: Faker::Number.between(1, 2),
```
Comment (igorkasyanchuk) [inline db/seeds.rb]: FIX, use tenants from DB or create 1-2-3 tenants before

File: db/seeds.rb [hunk 11]
```diff
@@ -19,3 +19,25 @@
...
+             created_at: rand(3.month.ago..Time.now),
+            })
+}
+User.all.each do |user|
+  rand(10).times {
+    random_date = rand(3.month.ago..Time.now)
+    user.visits.create(created_at: random_date, tenant_id: Faker::Number.between(1, 2))
```
Comment (igorkasyanchuk) [inline db/seeds.rb]: same here

File: app/controllers/account/dashboard_controller.rb [hunk 12]
```diff
@@ -2,5 +2,7 @@ class Account::DashboardController < ApplicationController
   before_action :authenticate_user!
   layout 'account'
 
-  def index; end
+  def index
+    render locals: { chart_value: current_companies }
```
Comment (romaslmd) [inline app/controllers/account/dashboard_controller.rb]: why not just assign to a instance variable instead?

File: db/seeds.rb [hunk 13]
```diff
@@ -19,3 +19,25 @@
 company = Company.first
 
 FactoryBot.create_list(:page, 20, company: company)
+
+200.times { Company.create({
```
Comment (romaslmd) [inline db/seeds.rb]: this is definitely not good..
use factories
als check `::create_list` from FactoryBot

File: db/seeds.rb [hunk 14]
```diff
@@ -19,3 +19,25 @@
...
+            User.create({
+             first_name: Faker::Name.first_name,
+             last_name: Faker::Name.last_name,
+             email: Faker::Internet.unique.email,
+             password: User::DEFAULT_PASSWORD,
+             tenant_id: Faker::Number.between(1, 2),
+             created_at: rand(3.month.ago..Time.now),
+            })
+}
+User.all.each do |user|
```
Comment (romaslmd) [inline db/seeds.rb]: this even worst
factories!

File: db/seeds.rb (hunk 11 above)
Comment (romaslmd) [inline db/seeds.rb]: `tenant_id` from faker? this is no test suite, this is seeds file

File: spec/controllers/charts_controller_spec.rb (hunk 9 above)
Comment (romaslmd) [reply spec/controllers/charts_controller_spec.rb]: `be_redirect` or `redirect_to`

Comment (romaslmd) [reply spec/controllers/charts_controller_spec.rb]: check for assigns at least

File: app/controllers/account/dashboard_controller.rb (hunk 12 above)
Comment (igorkasyanchuk) [reply app/controllers/account/dashboard_controller.rb]: agree with Roman. code is not good

File: app/controllers/account/charts_controller.rb [hunk 15]
```diff
@@ -0,0 +1,25 @@
+class Account::ChartsController < ApplicationController
+  before_action :authorize_admin!
+  def chart_for_user_registration
+    chart_data = current_tenant.users
+    render json: chart_data.group_by_week(:created_at, range: time_range).count
+  end
+
+  def chart_for_companies
```
Comment (romaslmd) [inline app/controllers/account/charts_controller.rb]: still inconsistent naming
try to keep naming in reatlike readable style..
as is now you have
```
GET /account/charts/charts_for_companies
                          ^^^^^^^^^^^^^^^^^^^^
                                  sounds?
mb something like this?
GET /account/charts/companies
```

File: spec/controllers/charts_controller_spec.rb [hunk 16]
```diff
@@ -0,0 +1,31 @@
+require 'rails_helper'
+
+RSpec.describe Account::ChartsController, type: :controller do
+  describe "GET requests to actions" do
```
Comment (romaslmd) [inline spec/controllers/charts_controller_spec.rb]: there is no such component, this is too general description

just wrap each metod into separate describe
also tests are a bit poor.. mb add some data validity expectations?

File: spec/factories/visits.rb [hunk 17]
```diff
@@ -0,0 +1,6 @@
+FactoryBot.define do
+  factory :visit, class: Visit do
+    user
+    created_at { rand(3.month.ago..Time.now) }
```
Comment (romaslmd) [inline spec/factories/visits.rb]: attribute first
associations after

File: spec/factories/companies.rb [hunk 18]
```diff
@@ -3,6 +3,7 @@
     user
     name { Faker::Company.name }
     domain { Faker::Internet.domain_name }
+    created_at { rand(3.month.ago..Time.now) }
```
Comment (romaslmd) [inline spec/factories/companies.rb]: dont override default behavior

use traits instead

File: app/controllers/account/charts_controller.rb (hunk 15 above)
Comment (igorkasyanchuk) [reply app/controllers/account/charts_controller.rb]: I would recomment to "chart_created_companites" or without "chart_" to be more clear what it returns

File: spec/controllers/charts_controller_spec.rb (hunk 16 above)
Comment (igorkasyanchuk) [reply spec/controllers/charts_controller_spec.rb]: it's also OK. you can remove this describe at all :)

File: spec/controllers/charts_controller_spec.rb [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+      get :chart_for_user_registration
+      expect(response).to have_http_status(200)
+      expect(response.content_type).to eq "application/json"
+    end
+    it "returns a 200 status code with json response" do
+      get :chart_for_companies
+      expect(response).to have_http_status(200)
+      expect(response.content_type).to eq "application/json"
+    end
+    it "returns a 200 status code with json response" do
```
Comment (igorkasyanchuk) [inline spec/controllers/charts_controller_spec.rb]: add empty line before each "it" for better readability

File: spec/factories/companies.rb (hunk 18 above)
Comment (igorkasyanchuk) [reply spec/factories/companies.rb]: yes

File: spec/factories/user.rb [hunk 20]
```diff
@@ -5,6 +5,7 @@
     sequence(:email) { |n| "#{n}#{Faker::Internet.email}" }
     tenant { build(:tenant) }
     password { Faker::Internet.password }
+    created_at { rand(3.month.ago..Time.now) }
```
Comment (igorkasyanchuk) [inline spec/factories/user.rb]: and here use trait.
something like "create(:user, :old_random_date)"

File: spec/factories/visits.rb (hunk 17 above)
Comment (igorkasyanchuk) [reply spec/factories/visits.rb]: actually it's ok :)  http://prntscr.com/i0apzu from official page

but use trait. 
I don't want to make this change global.

**Pull request merged**